Added mixed_precision support for preprocessing layers.

sebastian-sz opened the pull request:
# What does this PR do?
It was reported that some preprocessing layers do not run in mixed float16. This PR is an attempt to fix this.

Fixes #734 

This PR introduces two things:
1. Adds test for running in mixed precision for Keras-CV preprocessing layers.
2. Fix those layers that can be fixed from Python side.

### Affected layers:
Running (added in this PR) `keras_cv/layers/preprocessing/with_mixed_precision_test.py`, without changing the existing code, shows which layers are affected:
* GridMask
* CutMix
* Mosaic (throws some `OP_REQUIRES failed at functional_ops.cc:373` warning)
* RandomColorDegeneration 
* RandomCutout 
* RandomGaussianBlur
* RandomSharpness
* Solarization
* RandomColorJitter  
* RandomSaturation 

I managed to fix all, but two layers: RandomColorJitter and RandomSaturation. Both of these use `tf.image.adjust_saturation` which doesn't have CPU Float16 kernel registered. I believe this would have to be fixed from Tensorflow side.

The above layers should still work on the GPU though.

# Limitations
The majority of the problems were caused by explicitly casting to float32, which worked if the images were also in float32. I mostly added casting to `images.dtype` or fetching some Factor component with different dtype.

I am not sure, however, about running some computations in float16, as previously they were running in float32.
For example in: Mosaic and RandomGaussianBlur. I added a TODO so you could provide a comment whether you think this is ok or not.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?
@LukeWood @tanzhenyu @ianstenbit

ianstenbit commented on keras_cv/layers/preprocessing/random_color_degeneration.py:
```diff
@@ -55,7 +55,7 @@ def __init__(
 
     def get_random_transformation(self, **kwargs):
-        return self.factor()
+        return self.factor(dtype=kwargs["image"].dtype)
```
Everywhere where we're making this change -- instead of using `kwargs["image"]` let's just add `image=None` to the method signature

tanzhenyu replied:
+1

ianstenbit commented on keras_cv/layers/preprocessing/with_mixed_precision_test.py:
```diff
@@ -0,0 +1,136 @@
+class WithMixedPrecisionTest(tf.test.TestCase, parameterized.TestCase):
+    @parameterized.named_parameters(
+        *TEST_CONFIGURATIONS,
+        ("CutMix", preprocessing.CutMix, {}),
```
I think these can just be included in TEST_CONFIGURATIONS

sebastian-sz replied:
Added.

ianstenbit commented on keras_cv/layers/preprocessing/random_gaussian_blur.py:
```diff
@@ -94,12 +94,13 @@ def augment_segmentation_mask(self, segmentation_mask, transformation, **kwargs)
-            -tf.pow(x, 2.0) / (2.0 * tf.pow(tf.cast(factor, dtype=tf.float32), 2.0))
+            tf.range(-filter_size // 2 + 1, filter_size // 2 + 1), dtype=factor.dtype
         )
+        # TODO: are we ok with running this in FP16?
```
I would expect this to be okay in fp16, and unless there's some good evidence that this is bad, I don't think we should special case this.

tanzhenyu replied:
exponential computation might be a problem in fp16, how do we verify if it's ok to do this?

bhack replied:
In PyTorch vision they have test coverage for fp16:

https://github.com/pytorch/vision/blob/main/test/test_functional_tensor.py#L1097-L1146

sebastian-sz replied:
I inspected the output pixel values of `RandomGaussianBlur`. If we calculate the kernel in FP16 the results are quite off, compared to FP32 version (e.g. given value range of 0-255, FP32 will output 240 and FP16 will output 160). 

I will keep kernel calculation in FP32, as this fixes the issue. The blurring is still done in FP16 (after casting the kernel to adequate dtype).

sebastian-sz replied:
Moved (only) kernel calculation fo FP32.

ianstenbit reviewed (changes requested):
This is great! Thanks for the PR 💯

sebastian-sz commented:
@ianstenbit @tanzhenyu 
I updated the PR.  
* Removed `kwargs['image']` - use `self.compute_dtype` instead. I also added `self.compute_dtype` to most casts as I think it's more elegant (?).
* Moved `RandomGaussianBlur` kernel calculation to float32 (blurring, aka `tf.nn.depthwise_conv2d` still happens in float16).
* Added `dtype=self.compute_dtype` to `transform_value_range` functions, as it was force casting to float32 by default.

sebastian-sz commented on keras_cv/layers/preprocessing/random_gaussian_blur.py:
```diff
@@ -94,13 +98,12 @@ def augment_segmentation_mask(self, segmentation_mask, transformation, **kwargs)
-
         x = tf.cast(
-            tf.range(-filter_size // 2 + 1, filter_size // 2 + 1), dtype=factor.dtype
+            tf.range(-filter_size // 2 + 1, filter_size // 2 + 1), dtype=tf.float32
```
Added comment in `get_kernel` method.

ianstenbit commented:
/gcbrun
